File: learnLeetCode_remote/Day16/Part5/pickup_delivery.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 1359 — Count All Valid Pickup and Delivery Options
// Answer = prod_{i=1..n} i * (2*i - 1) mod 1e9+7

static const long long MOD = 1'000'000'007LL;

int countOrders(int n){
    long long ans = 1;
    for (int i=1;i<=n;++i){
        ans = (ans * i) % MOD;
        ans = (ans * (2LL*i - 1)) % MOD;
    }
    return (int)ans;
}

int main(){
    cout << countOrders(3) << "\n"; // 90
    return 0;
}

File: learnLeetCode_remote/Day16/Part5/pickup_delivery.py
# LC 1359 — Count All Valid Pickup and Delivery Options
# Formula: product over i=1..n of i * (2*i - 1) mod 1e9+7

MOD = 10**9 + 7

def countOrders(n: int) -> int:
    ans = 1
    for i in range(1, n+1):
        ans = (ans * i) % MOD         # place pickup i among previous pairs
        ans = (ans * (2*i - 1)) % MOD # place delivery i among (2*i - 1) slots
    return ans

if __name__ == "__main__":
    print(countOrders(3))  # 90
